daml-assistant/src/DA/Daml/Assistant/Version.hs

```diff
@@ -121,100 +126,151 @@ getDefaultSdkVersion damlPath = do
-            (pack . show $ getResponseStatus response)
-
-    versionsMap :: M.HashMap Text Text <-
-        fromRightM
-            (throwIO . assistantErrorBecause "Versions list from docs.daml.com does not contain valid JSON" . pack)
-            (eitherDecodeStrict' (getResponseBody response))
-
-    pure . sort $ mapMaybe (eitherToMaybe . parseVersion) (M.keys versionsMap)
-
--- | Get the list of available snapshot versions. This will fetch
--- https://docs.daml.com/snapshots.json and parse the obtained list
--- of versions.
-getAvailableSdkSnapshotVersions :: IO [SdkVersion]
-getAvailableSdkSnapshotVersions = wrapErr "Fetching list of available SDK snapshot versions" $ do
-    response <- requiredAny "HTTP connection to docs.daml.com failed" $ do
-        request <- parseRequest "GET http://docs.daml.com/snapshots.json"
-        httpBS request { responseTimeout = responseTimeoutMicro 2000000 }
-
-    when (getResponseStatusCode response /= 200) $ do
-        throwIO $ assistantErrorBecause
-            "Fetching list of available SDK snapshot versions from docs.daml.com failed"
-            (pack . show $ getResponseStatus response)
-
-    versionsMap :: M.HashMap Text Text <-
+isReleaseVersion :: SdkVersion -> Bool
+isReleaseVersion sdkVersion =
+    let v = unwrapSdkVersion sdkVersion
+    in
+    null (view V.release v) && null (view V.metadata v) && view V.major v > 0
+
+-- | Get the list of available snapshot versions. This will fetch all snapshot
+-- versions and then prune them into releases
+getAvailableSdkVersions :: UseCache -> IO ([SdkVersion], CacheAge)
+getAvailableSdkVersions useCache = do
+    (versions, cacheAge) <- wrapErr "Fetching list of available SDK versions" $ getAvailableSdkSnapshotVersions useCache
+    pure (extractVersionsFromSnapshots versions, cacheAge)
+
+extractVersionsFromSnapshots :: [SdkVersion] -> [SdkVersion]
+extractVersionsFromSnapshots snapshots =
+    let -- For grouping things by their major or minor version
+        distinguishBy :: Ord k => (a -> k) -> [a] -> M.Map k (NonEmpty.NonEmpty a)
+        distinguishBy f as = M.fromListWith (<>) [(f a, pure a) | a <- as]
+
+        -- Group versions by their major version number, filtering out snapshots
+        majorMap :: M.Map Int (NonEmpty.NonEmpty SdkVersion)
+        majorMap = distinguishBy (view V.major . unwrapSdkVersion) (filter isReleaseVersion snapshots)
+    in
+    case M.maxView majorMap of
+      Just (latestMajorVersions, withoutLatestMajor) ->
+        let -- For old majors, only the latest stable patch
+            oldMajors :: [SdkVersion]
+            oldMajors = map maximum (M.elems withoutLatestMajor)
+
+            latestMajorMinorVersions :: M.Map Int (NonEmpty.NonEmpty SdkVersion)
+            latestMajorMinorVersions =
+                distinguishBy (view V.minor . unwrapSdkVersion) (NonEmpty.toList latestMajorVersions)
+
+            -- For the most recent major version, output the latest minor version
```

Comment:
```suggestion
            -- For the most recent major version, output the latest minor versions
```
Is this correct? we want all the latest "patches" to the minor versions?

Reply:
I think we want:
- Filter out major 0.
- For all previous non-zero majors, only display the latest patch of the latest minor.
- For the current major, the latest patch of each minor.

But you should confirm that with whoever your product person is. The above may be too much.

Reply:
Pinged product on this :+1: